src/arcsf/data/tofu.py

```diff
@@ -0,0 +1,193 @@
+        elif not stratified:
+            num_forget = int(num_authors * q_per_author * forgotten_fact_fraction)
+
+    debug_dict = {
+        "author_count": num_authors,
+        "q_per_author": q_per_author,
+        "num_forget": num_forget,
+    }
+
+    # Get full list of indices
+    all_indices = list(range(all_data.num_rows))
+
+    # Get full list of authors
+    authors = list(range(num_authors))
+
+    # Conditionally create full list of questions
+    if granularity == "author" or not forget_random:
+        forget_questions = list(range(num_forget))
+
+    # create lists of retained and forgotten authors in all cases (sometimes not used)
+    retain_authors, forget_authors = train_test_split(
+        authors,
+        test_size=forgotten_author_fraction,
+        random_state=random_seed,
+    )
+
+    debug_dict["forget_author_numbers"] = forget_authors
+
+    if granularity == "question":
+        if forget_random:
+            if stratified:
+                # if stratified and random: select train-test split within each
+                # set of author questions (currently selects same indices
+                # TODO: Is this the behaviour we want?
+                forget_indices = []
+                for author in forget_authors:
+                    all_author_indices = np.arange(
+                        get_forget_indices(author, 0, q_per_author),
+                        get_forget_indices(author, 20, q_per_author),
+                    )
+                    _, author_forget_indices = train_test_split(
+                        all_author_indices,
+                        test_size=num_forget,
+                        random_state=random_seed,
+                    )
+                    forget_indices.append(author_forget_indices)
+                forget_indices = np.concatenate(forget_indices).tolist()
+                retain_indices = list(set(all_indices).difference(forget_indices))
+
+            # if not stratified we can apply train_test_split across entire dataset
+            else:
+                retain_indices, forget_indices = train_test_split(
+                    all_indices,
+                    test_size=num_forget,
+                    random_state=random_seed,
+                )
+                # need to redefine the forgotten authors now - can use np.floor
+                debug_dict["forget_author_numbers"] = np.floor(
+                    np.array(forget_indices) / 20
+                ).tolist()
+        # if not random then take first num_forget questions
+        # forget_questions contains the relevant indices for this
+        if not forget_random:
+            forget_indices = get_forget_indices(
+                forget_authors, forget_questions, q_per_author
+            )
+            retain_indices = list(set(all_indices).difference(forget_indices))
+
+    # if granularity == author we just use retain/forget authors to define our splits
+    if granularity == "author":
```

Comment:
```suggestion
    else:
```

or `elif granularity == "author"` (but question and author are the only options)

Reply:
changed